File: Solutions/876_Middle_of_the_Linked_List/876_Middle_of_the_Linked_List.py
#Third fastest
class Solution:
    def middleNode(self, head: ListNode) -> ListNode:
        slow, fast = head, head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        return slow

#Second Fastest
class Solution:
    def middleNode(self, head: ListNode) -> ListNode:
        slow = head
        fast = head
        while fast != None and fast.next != None:
            slow = slow.next
            fast = fast.next.next
        return slow 

#Fastest
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: ListNode) -> ListNode:
        fast_ptr = head
        slow_ptr = head
        while fast_ptr is not None:
            try:
                fast_ptr = fast_ptr.next.next
                slow_ptr = slow_ptr.next
            except:
                return slow_ptr
        return slow_ptr

File: Solutions/876_Middle_of_the_Linked_List/876_Middle_of_the_Linked_List.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode MiddleNode(ListNode head) {
        ListNode midPrev = null;
        while(head != null && head.next != null) {
            midPrev = (midPrev == null) ? head : midPrev.next;
            head = head.next.next;
        }
        ListNode mid = midPrev == null ? head : midPrev.next;
        return mid;
    }
}

//Third fastest
public class Solution {
    public ListNode MiddleNode(ListNode head) {
        ListNode fast = head;
        ListNode slow = head;
        while (fast != null && fast.next != null)
        { 
            fast = fast.next.next;
            slow = slow.next;
        }
        return slow;
    }

//Second fastest
public class Solution {
    public ListNode MiddleNode(ListNode head) {        
        var s = head;
        var f = head;
        
        while(f != null && f.next != null){
            s = s.next;
            f = f.next.next;
        }
        
        return s;
    }
}

//Fastest
public class Solution {
    public ListNode MiddleNode(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;
        while(fast != null && fast.next != null)
        {
            slow = slow.next;
            fast = fast.next.next;
        }
        return slow ;
        
        
    }
}